### PR TITLE
use iglp to improve dim256 fmha fwd in qr_ks_vs pipeline

### PR DESCRIPTION
**Pull request by LJ-underdog**
## Proposed changes

use iglp in qr_ks_vs pipeline and increase k_lds_padding to improve dim256 fmha fwd .

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

### CODE REVIEW

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 1]
```diff
@@ -281,7 +281,20 @@ struct BlockFmhaPipelineQRKSVS
         index_t i_total_loops      = 0;
         constexpr index_t k0_loops = kQKHeaddim / kK0;
         constexpr index_t k1_loops = kN0 / kK1;
-
+        auto group_barrier_seq = (kQKHeaddim == 256)
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: extract this as another class template. Of simpliy declare a lambda function and wrap code in the if constexpr() clause.
```c++
// method 1
template <bool Eanble>
class Scheduler
{
    static constexpr void schedule_gemm0() { }
};

template <>
class Scheduler<true>
{
    static constexpr void schedule_gemm0() { }
};

Scheduler<kQKHeaddim == 256> scheduler;
scheduler.schedule_gemm0();

// method 2
auto schedule_gemm0 = [] {
   if constexpr(kQKHeaddim == 256) {
      // static_for<>...
   }
};
schedule_gemm0();
```

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 2]
```diff
@@ -281,7 +281,20 @@ struct BlockFmhaPipelineQRKSVS
...
+    ? []() {
+        static_for<0, 2, 1>{}([&](auto) {
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: instead of using magic number, define something like:
```cpp
constexpr int num_mfma_insts = /* compute from the tile size & warp gemm size */;
static_assert(num_mfma_insts % 8 == 0); // we have 8 mfma in static_for<>
static_for<0, num_mfma_insts / 8, 1>{}([](auto) { // no need capture [&] here
});
```

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 3]
```diff
@@ -281,7 +281,20 @@ struct BlockFmhaPipelineQRKSVS
...
+        __builtin_amdgcn_sched_group_barrier(0x100, 2, 0); // DS read
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: consider removing the magic numbers: `0x100`, `0x008`

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qx_ks_vs_custom_policy.hpp [hunk 4]
```diff
@@ -500,7 +500,7 @@ struct BlockFmhaPipelineQXKSVSCustomPolicy : BlockFmhaPipelineQXCustomPolicy<QLo
 
         constexpr auto k_lds_block_desc_0 = make_naive_tensor_descriptor(
             make_tuple(number<kKPerBlock / kKPack>{}, number<kNPerBlock>{}, number<kKPack>{}),
-            make_tuple(number<(kNPerBlock + 1) * kKPack>{}, number<kKPack>{}, number<1>{}),
+            make_tuple(number<(kNPerBlock + 2) * kKPack>{}, number<kKPack>{}, number<1>{}),
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qx_ks_vs_custom_policy.hpp]: Is this change always good for performance? Have you tried more workloads?

Comment (LJ-underdog) [reply include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qx_ks_vs_custom_policy.hpp]: I found an error when use padding=2 ,so I remove this change.
`./bin/tile_example_fmha_fwd -prec=fp16 -mode=1 -b=2 -h=1 -d=256 -d_v=24 -s=3 -s_k=99 -bias=n -p_drop=0.2 -lse=0 -iperm=0 -operm=0 -mask=2 -num_splits=1 -page_block_size=0 -cache_batch_idx=0 -kname=1 -v=1 -warmup=0 -repeat=1
[fp16|group|bshd] b:2, h:1/1, s:3/99, d:256/24, scale_s:0.0625, bias:n, p_drop:0.2, lse:0, squant:0, mask:b(-1:0), v:r, fmha_fwd_d256_fp16_group_b128x128x32x256x32x256_r4x1x1_r4x1x1_w32x32x16_w32x32x16_qr_vr_psskddv_nlogits_nbias_mask_nlse_dropout_nskip_nsquant_ntrload, 1.018 ms, 0.00 TFlops, 0.11 GB/s
OUT Error: Incorrect results!        out[34] != ref[34]: 0.482666 != 0.4804688
OUT Error: Incorrect results!        out[45] != ref[45]: 0.5292969 != 0.5273438
OUT Error: Incorrect results!        out[52] != ref[52]: 0.4194336 != 0.4174805
OUT Error: Incorrect results!        out[61] != ref[61]: 0.4384766 != 0.4365234
max err: 0.002197266, number of errors: 7, 9.722222% wrong values
OUT mismatch found at batch: 1
	seqlen_q: 3
	seqlen_k: 99
	seqstart_q: [0, 3, 6]
	seqstart_k: [0, 99, 198]
, valid:n`

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 5]
```diff
@@ -58,6 +58,19 @@ struct BlockFmhaPipelineQRKSVS
...
 
+    using Gemm_0 = remove_cvref_t<decltype(Policy::template GetQKBlockGemm<Problem>())>;
+    static constexpr auto WG0_config = Gemm_0::Policy::template GetWarpGemmMWarpNWarp<Problem>();
+    using WG0                        = remove_cvref_t<decltype(WG0_config.template at<0>())>;
+    static constexpr index_t MWarp0  = WG0_config.template at<1>();
+    static constexpr index_t NWarp0  = WG0_config.template at<2>();
+    static constexpr index_t WG0_kM  = WG0 ::WarpGemmAttribute::Impl::kM;
+    static constexpr index_t WG0_kN  = WG0 ::WarpGemmAttribute::Impl::kN;
+    static constexpr index_t WG0_kK  = WG0 ::WarpGemmAttribute::Impl::kK;
+    static constexpr uint32_t DS_READ_BARRIER = 0x100; // Barrier for DS (data share) read
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: suggest to rename `DS_READ_BARRIER` as `SCHED_MARK_DS_READ` to align with the enum class definition [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp#L65-L81)

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 6]
```diff
@@ -58,6 +58,19 @@ struct BlockFmhaPipelineQRKSVS
     static constexpr bool kStoreLSE         = Problem::kStoreLSE;
     static constexpr bool kHasDropout       = Problem::kHasDropout;
 
+    using Gemm_0 = remove_cvref_t<decltype(Policy::template GetQKBlockGemm<Problem>())>;
+    static constexpr auto WG0_config = Gemm_0::Policy::template GetWarpGemmMWarpNWarp<Problem>();
+    using WG0                        = remove_cvref_t<decltype(WG0_config.template at<0>())>;
+    static constexpr index_t MWarp0  = WG0_config.template at<1>();
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: suggest to unify naming style, we have lots of styles here `Gemm_0`, `WG0`, `WG0_kM`

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp [hunk 7]
```diff
@@ -58,6 +58,19 @@ struct BlockFmhaPipelineQRKSVS
...
+    static constexpr index_t NWarp0  = WG0_config.template at<2>();
+    static constexpr index_t WG0_kM  = WG0 ::WarpGemmAttribute::Impl::kM;
+    static constexpr index_t WG0_kN  = WG0 ::WarpGemmAttribute::Impl::kN;
+    static constexpr index_t WG0_kK  = WG0 ::WarpGemmAttribute::Impl::kK;
+    static constexpr uint32_t DS_READ_BARRIER = 0x100; // Barrier for DS (data share) read
+    static constexpr uint32_t MFMA_BARRIER = 0x008; // Barrier for MFMA (matrix multiply-accumulate)
+    static constexpr int num_mfma_insts =
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_qr_ks_vs.hpp]: use same naming style as other constexpr variables

**Comment by LJ-underdog** [comment]
fmha test passed: http://micimaster.amd.com/blue/organizations/jenkins/MLLIBS%2Fcomposable_kernel/detail/linjun%2Fdim256_optimization/14/pipeline/248/

**Comment by poyenc** [comment]
LGTM. We will need template traits to get block gemm properties. So that we don't have to declare them the pipeline. You can add this in separate PR.

**Review by poyenc — APPROVED** [review]
